Improve repo level ordering tf

shivdeep-singh-ibm opened the pull request:
## Why are these changes needed?


## Related issue number (if any).
https://github.com/IBM/data-prep-kit/issues/421
https://github.com/IBM/data-prep-kit/issues/420

Param-S commented on transforms/code/repo_level_ordering/README.md:
```diff
@@ -0,0 +1,8 @@
+# Repo Level Order Transform 
+The repo level order transforms serves as a simple exemplar to demonstrate the development
```
Description is not correctly reflecting the functionality. It is basically transforms 1 parquet into multiple parquet output files based on the no. of unique repos found.

shivdeep-singh-ibm replied:
updated.

daw3rd commented on data-processing-lib/ray/src/data_processing_ray/runtime/ray/transform_orchestrator.py:
```diff
@@ -131,6 +131,10 @@ def orchestrate(
+
+    # Cleanup processors and actor pool
+    del processors
+    del processors_pool
```
@blublinsky did we decide these belong in the transform itself?

daw3rd commented on examples/notebooks/code/launch.sh:
```diff
@@ -19,6 +19,7 @@ export PYTHONPATH=$PYTHONPATH:$REPO_ROOT/transforms/universal/doc_id/ray/src:$RE
 export PYTHONPATH=$PYTHONPATH:$REPO_ROOT/transforms/universal/tokenization/ray/src:$REPO_ROOT/transforms/universal/tokenization/python/src
 export PYTHONPATH=$PYTHONPATH:$REPO_ROOT/tools/ingest2parquet/src/
 export PYTHONPATH=$PYTHONPATH:$REPO_ROOT/transforms/code/ingest_2_parquet/ray/src:$REPO_ROOT/transforms/code/ingest_2_parquet/python/src
```
this is now called code2parquet

shivdeep-singh-ibm replied:
fixed

daw3rd commented on transforms/code/repo_level_ordering/README.md:
```diff
@@ -0,0 +1,8 @@
+# Repo Level Order Transform 
+The repo level order transforms serves as a simple exemplar to demonstrate the development
```
incorrect documentation

shivdeep-singh-ibm replied:
fixed

daw3rd commented on transforms/code/repo_level_ordering/ray/Makefile:
```diff
@@ -0,0 +1,51 @@
+set-versions: 
+	$(MAKE) TRANSFORM_PYTHON_VERSION=dummy TOML_VERSION=$(REPO_LVL_ORDER_RAY_VERSION) .transforms.set-versions 
+        
+build-dist:: set-versions .defaults.build-dist 
```
you can/should remove set-versions here.

shivdeep-singh-ibm replied:
fixed.

daw3rd commented on transforms/code/repo_level_ordering/ray/README.md:
```diff
@@ -0,0 +1,159 @@
+c) content. This transform supports searching the repo across the whole dataset and writing the files of a repo as a single 
+parquet file. 
+
+The transform gives the option to write the repo to file in the following ways.
```
"write the repo to a file"? I guess you're ordering rows of the input parquet file?  If so, you should say this.

shivdeep-singh-ibm replied:
fixed.

daw3rd commented on transforms/code/repo_level_ordering/ray/README.md:
```diff
@@ -0,0 +1,159 @@
+
+## Configuration and command line Options
+
+repo_level_order configuration and command line options are the same as for the base python transform. 
```
Uhmm, there is no python implementation.

shivdeep-singh-ibm replied:
changes in another PR, they don't show up here.

daw3rd commented on transforms/code/repo_level_ordering/ray/pyproject.toml:
```diff
@@ -0,0 +1,51 @@
+    { name = "Shanmukha Guttula", email = "shagutt1@in.ibm.com" },
+]
+dependencies = [
+    "data-prep-toolkit-ray==0.2.0",
```
this should be 0.2.1.dev0

shivdeep-singh-ibm replied:
fixed.

daw3rd commented on transforms/code/repo_level_ordering/ray/README.md:
````diff
@@ -0,0 +1,159 @@
+       --repo_lvl_sorting_enabled \
+       --repo_lvl_sorting_algo SORT_SEMANTIC \
+       --repo_lvl_output_by_langs   
+```
````
You should add the piece that other readmes have (at the end) about running as an image.

shivdeep-singh-ibm replied:
readme handled in another PR, changes won't show here.

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -0,0 +1,396 @@
+
+short_name = "repo_lvl"
+cli_prefix = f"{short_name}_"
+pwd_key = "pwd"
```
suspect you don't really need this

shivdeep-singh-ibm replied:
removed

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -0,0 +1,396 @@
+
+store_type_key = "store_type"
+store_dir_key = "store_backend_dir"
+store_s3_secret_key = "other_s3_secret"
```
many of these are not used.

shivdeep-singh-ibm replied:
removed

daw3rd reviewed (commented):
might be nice to call this `repo_order` instead, just to shorten things.  Then the shortname could match the dir name.

daw3rd commented on transforms/code/repo_level_ordering/ray/test/test_repo_level_order.py:
an improvement, but don't you need a transform test?

daw3rd replied:
and, by convention the file with this name contains the "transform" test that extends AbstractTableTransformTest.  Maybe these should be in another file?

daw3rd commented on examples/notebooks/code/demo_with_launcher.ipynb:
File references ingest2parquet, when code2parquet is being used.

shivdeep-singh-ibm replied:
done.

daw3rd commented on transforms/code/repo_level_ordering/ray/README.md:
You discuss the sorting and other functionality, but nothing on the configuration keys to control that functionality.

shivdeep-singh-ibm replied:
Done.

daw3rd replied:
This readme needs lots of work.  Please consider following pdf2parquet as an example, https://github.com/IBM/data-prep-kit/blob/dev/transforms/language/pdf2parquet/python/README.md#parameters.

daw3rd replied:
Still needs work

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -90,29 +75,54 @@ def __init__(self, config: dict[str, Any]):
-        a dictionary of execution statistics - arbitrary dictionary
-        This implementation makes no modifications so effectively implements a copy of the
-        input parquet to the output folder, without modification.
+        This step doesn't prepare data to write tables, this step is used to do groupby
```
Don't say what it doesn't do.  say what it does do :)

shivdeep-singh-ibm replied:
ok. Done.

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -90,29 +75,54 @@ def __init__(self, config: dict[str, Any]):
         self.grouping_column = config.get(grouping_column_key)
```
You need to implement defaults here OR throw exception for missing required configuration keys. Alternatively, since it seems this transform can only run in the Ray runtime, a check for that condition would be useful.  What we don't want is someone trying to run the transform outside of the runtime, but if they do, they should get an intelligent message.

shivdeep-singh-ibm replied:
done

daw3rd commented on transforms/code/repo_level_ordering/ray/test/test_transform.py:
```diff
@@ -0,0 +1,112 @@
+    cleanup_output(output_folder)
+    import tempfile
+
+    with tempfile.TemporaryDirectory() as store_backend_dir:
```
Why not implement these tests as AbstractTransformLauncherTest-base fixtures (extending your other test with additional fixtures)?  Would require defining a new test-data/expected output directory tree, but that would be the more comprehensive test, especially since it does a more rigourous validation of the output parquet.

shivdeep-singh-ibm replied:
This is simple test to check number of output files and folders. Yes, I agree more `AbstractTransformLauncherTest` fixtures can be added,

daw3rd replied:
checking output files and folders is what `AbstractTransformLauncherTest` does.  can you switch to that?  Or at least add similar tests using that framework?  I think it may be more comprehensive.

daw3rd replied:
Could use work here, but there are tests still.

Param-S commented on examples/notebooks/code/demo_with_launcher.ipynb:
```diff
@@ -90,10 +78,19 @@
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
```
I am just thinking whether we should include the output cells with output text or not. IMHO we should not include it. If we want to include then it should include an example run for a specific test dataset available in our repo.

shivdeep-singh-ibm replied:
Yes, the output is needed so that user can see the sample output from the notebook.

Param-S commented on transforms/code/repo_level_ordering/ray/src/dpk_repo_level_order/internal/repo_level_wrappers.py:
```diff
@@ -19,17 +19,23 @@
+def semantic_sort_normalised(df, logger, title_column_name, language_column_name):
     check_and_update_title(df)
-    return sort_sem(df, logger, title_column_name)
+    return sort_sem(df, logger, title_column_name, language_column_name)
```
here positional arguments notation used and at line no. 24 used keyword arguments notation for the same function invocation. can we have one invocation convention across the code. or is there any specific reason used different convention

shivdeep-singh-ibm replied:
corrected.

Param-S commented on transforms/code/repo_level_ordering/ray/src/dpk_repo_level_order/internal/store/ray_store.py:
```diff
@@ -83,10 +121,7 @@ def get(self, key):
-                    "Error: %s. Ignoring the error",
-                    e,
-                )
+                print(f"Error: {e}. Ignoring the error")
```
at 3 different places(line no. 92, 110 & 124) used the same string format. This might cause difficultly if we want to locate the failing point from the logs. can we have some uniqe keyword  in the string for each print statements.

shivdeep-singh-ibm replied:
corrected.

Param-S reviewed (commented):
Few minor suggestions for code readability. other than this the PR looks fine to me

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -300,7 +352,13 @@ def add_input_params(self, parser: ArgumentParser) -> None:
+            f"--{cli_prefix}{language_column_key}",
+            type=str,
+            default="language",
             help="The name of the column which has repo name.",
```
looks like copied help text

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -328,7 +386,8 @@ def add_input_params(self, parser: ArgumentParser) -> None:
-            action="store_true",
+            default=False,
+            type=bool,
             help="Enables sorting of output.",
```
sorting by what?

daw3rd commented on examples/notebooks/code/launch.sh:
```diff
@@ -18,7 +18,6 @@ export PYTHONPATH=$PYTHONPATH:$REPO_ROOT/transforms/universal/filter/ray/src:$RE
 export PYTHONPATH=$PYTHONPATH:$REPO_ROOT/transforms/universal/doc_id/ray/src:$REPO_ROOT/transforms/universal/doc_id/python/src
 export PYTHONPATH=$PYTHONPATH:$REPO_ROOT/transforms/universal/tokenization/ray/src:$REPO_ROOT/transforms/universal/tokenization/python/src
 export PYTHONPATH=$PYTHONPATH:$REPO_ROOT/tools/ingest2parquet/src/
```
the ingest2parquet tool is deprecated.  can you use code2parquet transform instead?

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -87,49 +90,94 @@ def __init__(self, config: dict[str, Any]):
         self.logger = get_logger(__name__)
         super().__init__(params)
 
         self.stage_one_only = self.params["stage_one_only"]
```
why not a constant here for the key, like the others

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -87,49 +90,94 @@ def __init__(self, config: dict[str, Any]):
+
         self.output_by_langs = self.params[output_by_langs_key]
         self.combine_rows = self.params[output_superrows_key]
         self.ray_workers = self.params["ray_workers"]
```
constant?

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -61,23 +51,36 @@
+                                     {'access_key', 'secret_key', 'url'},
+                                     None if store is ray or local
+                           store_pool_key: A list of actors if store_type_value_ray, otherwise None
+                           store_ray_cpus_key: Num of cpus used, if store_type_value_ray
```
i don't see these referenced in the initializer.  You should probably check to be sure they are present there.  This will help keep someone from trying to run the transform outside of ray too.

daw3rd replied:
nice to have, but not mandatory.

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -61,23 +51,36 @@
+                                        store_type_value_local for single node if cpus are less.
+                           store_backend_dir - A path for
+                                        store types [ store_type_value_s3, store_type_value_local]
+                           s3_creds: A dictionary for S3 creds if using store_type_value_s3.
```
s3_creds does not seem to be used.  maybe this whole doc section needs checking?

daw3rd commented on transforms/code/repo_level_ordering/ray/README.md:
The formatting of the a), b), c), d) items is wrapping and not showing up as individual bullets.

daw3rd replied:
and somewhere in here is a reference to the "base python transform"

daw3rd commented on transforms/code/repo_level_ordering/ray/src/repo_level_order_transform.py:
```diff
@@ -293,16 +350,22 @@ def add_input_params(self, parser: ArgumentParser) -> None:
+        parser.add_argument(
+            f"--{cli_prefix}{language_column_key}",
+            type=str,
+            default="language",
```
default should be in a variable, for consistency, like the others.

daw3rd replied:
really applies to the bool args as well.  i'm sure this seems like busy work, but its much nicer to be able to browse the top of the file for this type of thing than dig through 100s of lines of code to find the 1 line that sets the default.

Param-S commented on transforms/code/repo_level_ordering/kfp_ray/repo_level_order_wf.py:
```diff
@@ -0,0 +1,222 @@
+    runtime_pipeline_id: str,
+    runtime_job_id: str,
+    runtime_code_location: str,
+    cq_contents_column_name: str,
```
i think, these parameters related to code quality which need to be removed from here.

Param-S commented on transforms/code/repo_level_ordering/ray/src/dpk_repo_level_order/internal/repo_level_wrappers.py:
```diff
@@ -12,24 +12,33 @@
 
 
-def semantic_sort_normalised(df, logger, title_column_name):
+def semantic_sort_normalised(df, logger, title_column_name, language_column_name):
```
instead of having two functions(`semantic_sort` and `semantic_sort_normalised`, can we combine them into one by adding one additional optional parameter "normalize_title" with default value of `False`

```
semantic_sort(df, logger, title_column_name, language_column_name, normalize_title=False):
```

Param-S commented on transforms/code/repo_level_ordering/ray/src/dpk_repo_level_order/internal/repo_level_wrappers.py:
```diff
@@ -38,50 +47,72 @@ def get_sorting_func(
+            )
+        except FunctionTimedOut as e:
+            logger.error(
+                f"We got an exception while sorting [{file_name}].\n Exception: {e.__class__.__name__}.\n Falling back to default sort_by_path"
```
can we rephrase this error msg like: "Exception while sorting [{file_name}].\n Exception: {e.__class__.__name__}.\n Falling back to default sort_by_path"

Param-S commented on transforms/code/repo_level_ordering/ray/src/dpk_repo_level_order/internal/repo_level_wrappers.py:
```diff
@@ -38,50 +47,72 @@ def get_sorting_func(
+            sorted_df = sort_by_path(df=df, logger=logger, title_column_name=title_column_name)
+        except Exception as e:
+            logger.error(
+                f"We got an exception while sorting [{file_name}].\n Exception: {e.__class__.__name__}.\n Falling back to default sort_by_path"
```
same here

Param-S commented on transforms/code/repo_level_ordering/ray/src/dpk_repo_level_order/internal/repo_level_wrappers.py:
```diff
@@ -38,50 +47,72 @@ def get_sorting_func(
-    import random
+        """
+        lang_name = get_dominant_language_repo_packing(table, language_column_name, title_column_name)
+        if lang_name.lower in ["c#"]:
```
why do we need this special condition for one language.

Param-S commented on transforms/code/repo_level_ordering/ray/src/dpk_repo_level_order/internal/repo_level_wrappers.py:
```diff
@@ -38,50 +47,72 @@ def get_sorting_func(
+
+    super_row = table.column("contents").to_pylist()
+    repo_doc_ids = table.column("document_id").to_pylist()
+    print(table.column_names)
```
do we need to retain this print statement

Param-S commented on transforms/code/repo_level_ordering/ray/src/dpk_repo_level_order/internal/repo_level_wrappers.py:
```diff
@@ -38,50 +47,72 @@ def get_sorting_func(
 
 
-def superrow_table(table: pa.Table, repo_column_name: str) -> pa.Table:
+def superrow_table(table: pa.Table, repo_column_name: str, language_column_name="language") -> pa.Table:
```
can we have uniform default value for language_column_name variable: 

In other function, it is set to 

```
def get_dominant_language_repo_packing(table, language_column="programming_language", title_column="title"):
```

Param-S commented on transforms/code/repo_level_ordering/ray/src/dpk_repo_level_order/internal/repo_level_wrappers.py:
```diff
@@ -108,15 +139,15 @@ def superrow_table(table: pa.Table, repo_column_name: str) -> pa.Table:
 
 
-def get_transforming_func(sorting_func=None, superrows_func=None, filename_func=None):
+def get_transforming_func(sorting_func=None, superrows_func=None, filename_func=None, language_column_name="language"):
```
same here: can we make it uniform default value

daw3rd commented on transforms/code/repo_level_ordering/ray/test/test_internal_wrappers.py:
```diff
@@ -0,0 +1,129 @@
+)
+
+
+class PassThroughTransform(AbstractTableTransform):
```
test_support package in the main python lib includes a NoopTransform.  If this meets your needs, may be use it instead.

daw3rd replied:
nice to have, but not mandatory

Param-S reviewed (approved):
LGTM

Param-S commented:
@shivdeep-singh-ibm can you resolve the conflicts, we can merge this PR once it is resolved

Param-S commented:
@shivdeep-singh-ibm can you please check the testcase, 2 failures reported https://github.com/IBM/data-prep-kit/actions/runs/10416381352/job/28848510989?pr=434

Param-S commented:
@shivdeep-singh-ibm can you please resolve the merge conflicts

shivdeep-singh-ibm committed:
Update repo level ordering transform

    - change flags to cli args
    - add tests
    - add batching while storing groups
    - optimise reading files in GroupByRepo

Signed-off-by: Shivdeep Singh <Shivdeep.Singh@ibm.com>

Param-S reviewed (approved):
lgtm
